server/lib/collectivelib.ts

```diff
@@ -290,63 +290,76 @@ export async function isCollectiveDeletable(collective) {
+          AND other_members."deletedAt" IS NULL
+        WHERE m.role = 'ADMIN'
+        AND m."MemberCollectiveId" = :CollectiveId
+        AND m."deletedAt" IS NULL
+        AND c."deletedAt" IS NULL
+        AND other_members.id IS NULL
+      ) AS "isLastAdminOfAnyCollective"
+    `,
+      { plain: true, replacements: { CollectiveId: collective.id } },
+    );
+
+    if (isLastAdminOfAnyCollective) {
+      return false;
     }
-  }
-
-  const transactionCount = await models.Transaction.count({
-    where: {
-      [Op.or]: [{ CollectiveId: collective.id }, { FromCollectiveId: collective.id }],
-    },
-  });
-
-  const orderCount = await models.Order.count({
-    where: {
-      [Op.or]: [{ CollectiveId: collective.id }, { FromCollectiveId: collective.id }],
-      status: ['PAID', 'ACTIVE', 'CANCELLED'],
-    },
-  });
 
-  let expenseCount;
-  if (user) {
-    expenseCount = await models.Expense.count({
-      where: {
-        [Op.or]: [{ CollectiveId: collective.id }, { FromCollectiveId: collective.id }, { UserId: user.id }],
-        status: ['PAID', 'PROCESSING', 'SCHEDULED_FOR_PAYMENT'],
-      },
-    });
-  } else {
-    expenseCount = await models.Expense.count({
-      where: {
-        [Op.or]: [{ CollectiveId: collective.id }, { FromCollectiveId: collective.id }],
-        status: ['PAID', 'PROCESSING', 'SCHEDULED_FOR_PAYMENT'],
-      },
-    });
+    user = await models.User.findOne({ where: { CollectiveId: collective.id } });
   }
 
-  const childrenCount = await models.Collective.count({
-    where: { ParentCollectiveId: collective.id, type: { [Op.in]: [EVENT, PROJECT] } },
-  });
-
-  if (transactionCount > 0 || orderCount > 0 || expenseCount > 0 || childrenCount > 0) {
-    return false;
-  }
+  const { hasUndeletableData } = await sequelize.query(
+    `
+    SELECT (
+      -- Children
+      EXISTS (
+        SELECT 1 FROM "Collectives"
+        WHERE "ParentCollectiveId" = :CollectiveId
+        AND "deletedAt" IS NULL
+      )
+      -- Expenses
+      OR EXISTS (
+        SELECT 1 FROM "Expenses"
+        WHERE (
+          "CollectiveId" = :CollectiveId
+          OR "FromCollectiveId" = :CollectiveId
+          ${user ? `OR "UserId" = :UserId` : ''}
+        )
+        AND "deletedAt" IS NULL
+      )
+      -- Orders
+      OR EXISTS (
+        SELECT 1 FROM "Orders"
+        WHERE ("CollectiveId" = :CollectiveId OR "FromCollectiveId" = :CollectiveId)
+        AND "deletedAt" IS NULL
+        AND status IN ('PAID', 'ACTIVE', 'CANCELLED')
+      )
+      -- Transactions
+      OR EXISTS (
+        SELECT 1 FROM "Transactions"
+        WHERE ("CollectiveId" = :CollectiveId OR "FromCollectiveId" = :CollectiveId)
+        AND "deletedAt" IS NULL
+      )
+    ) AS "hasUndeletableData"
+  `,
+    {
+      plain: true,
+      replacements: { CollectiveId: collective.id, UserId: user?.id },
+    },
+  );
 
-  return true;
+  return !hasUndeletableData;
```

Comment:
This query brings a few optimizations compared to the previous:
- We were fetching all the data, regardless of the previous results. i.e. we were counting all expenses even if we detected transactions before. By relying on SQL's ORs, we're making sure we stop querying data as soon as we find something that can prevent the deletion.
- The query order matters: `Transactions` are heavier to fetch because it's a bigger table. By fetching them at the end, we're making sure that we first look at lighter tables.

Explain analyze with OSC (look at the "never executed"):
```
Result  (cost=24.18..24.18 rows=1 width=1) (actual time=0.015..0.016 rows=1 loops=1)
  InitPlan 1 (returns $0)
"    ->  Index Scan using ""ParentCollectiveId"" on ""Collectives""  (cost=0.08..11.28 rows=5 width=0) (actual time=0.013..0.013 rows=1 loops=1)"
"          Index Cond: (""ParentCollectiveId"" = 11004)"
"          Filter: (""deletedAt"" IS NULL)"
  InitPlan 2 (returns $1)
"    ->  Bitmap Heap Scan on ""Expenses""  (cost=7.09..1000.20 rows=609 width=0) (never executed)"
"          Recheck Cond: (((""CollectiveId"" = 11004) AND (""deletedAt"" IS NULL)) OR ((""FromCollectiveId"" = 11004) AND (""deletedAt"" IS NULL)))"
          ->  BitmapOr  (cost=7.09..7.09 rows=609 width=0) (never executed)
                ->  Bitmap Index Scan on expenses_collective_id  (cost=0.00..4.96 rows=602 width=0) (never executed)
"                      Index Cond: (""CollectiveId"" = 11004)"
                ->  Bitmap Index Scan on expenses_from_collective_id  (cost=0.00..2.07 rows=7 width=0) (never executed)
"                      Index Cond: (""FromCollectiveId"" = 11004)"
  InitPlan 3 (returns $2)
"    ->  Bitmap Heap Scan on ""Orders""  (cost=6.81..812.36 rows=132 width=0) (never executed)"
"          Recheck Cond: (((""CollectiveId"" = 11004) AND (""deletedAt"" IS NULL)) OR ((""FromCollectiveId"" = 11004) AND (""deletedAt"" IS NULL)))"
          Filter: ((status)::text = ANY ('{PAID,ACTIVE,CANCELLED}'::text[]))
          ->  BitmapOr  (cost=6.81..6.81 rows=416 width=0) (never executed)
                ->  Bitmap Index Scan on orders_collective_id  (cost=0.00..4.70 rows=412 width=0) (never executed)
"                      Index Cond: (""CollectiveId"" = 11004)"
                ->  Bitmap Index Scan on orders_from_collective_id  (cost=0.00..2.09 rows=4 width=0) (never executed)
"                      Index Cond: (""FromCollectiveId"" = 11004)"
  InitPlan 4 (returns $3)
"    ->  Seq Scan on ""Transactions""  (cost=0.00..183390.16 rows=830198 width=0) (never executed)"
"          Filter: ((""deletedAt"" IS NULL) AND ((""CollectiveId"" = 11004) OR (""FromCollectiveId"" = 11004)))"
Planning Time: 0.440 ms
Execution Time: 0.060 ms
```